### ISSUE TITLE
Customization for jupyter notebook

### ISSUE THREAD

**Issue by Unknwon** [gogits/gogs#4097]
Currently, syntax highlight is [hard-coded for Python](https://github.com/gogits/gogs/blob/master/templates/repo/view_file.tmpl#L51), and could cause problem when using notebook for [other languages](https://github.com/jupyter/jupyter/wiki/Jupyter-kernels).

Since all languages' notebook is using file extension `.ipynb`, it is impossible to set highlight language based on it. We could let highlightJS dynamically decide itself, but it is [considered unreliable](https://github.com/gogits/gogs/issues/3715#issuecomment-278447705).

Thus, to have a repo-based customization/configuration for notebook is needed, allow repository owner to decide which language to highlighted as.

Here is the basic proposal:

Based on the fact we already uses `.gogs` directory for issue/pull request templates, we can add another file `ipynb.ini` or `ipynb.json` (format to be determined, but prefer `.ini` like `.editorconfig`).

**Comment by prinsherbert** [comment]
I would disagree on the configuration idea. It has some code overhead and notebook syntax highlighting might not work out of the box. As I stated in https://github.com/gogits/gogs/issues/3715 , the json of the notebook contains information about the language in the ipynb-notebooks, on document-level and/or on cell-level. We should map those values to appropriate syntax highlighting classes and disable syntax highlighting for unknown languages (or detect it).

**Comment by Unknwon** [comment]
@prinsherbert I'm not understand very well here, is it possible to have multiple language in a single ipynb file?

**Comment by oroszl** [comment]
it is possible to have a notebook with different languages e.g. with cell magics but I am not entirely sure if jupyter itself handles highlighting properly in that case.

**Comment by Unknwon** [comment]
>it is possible to have a notebook with different languages

This probably too complicated cases to handle...

But in general, you don't want R code to be forced highlighted as Python, do you?

**Comment by oroszl** [comment]
No I don't think so, to be honest it would be nice to have the option to let highlights.js decide.

**Comment by Unknwon** [comment]
OK... maybe I should just say... fork Gogs and make it whatever you want..

**Comment by oroszl** [comment]
:) fair point

**Comment by khimaros** [comment]
Perhaps what @prinsherbert is suggesting here is that the syntax highlighting language can be selected based on metadata in the .ipynb file itself.

Here's a concrete example:

```
{
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "pygments_lexer": "ipython3",
  }
 },
}
```

As you can see the pygments lexer and language are specified as part of the ipynb metadata. This could be passed as a hint to highlightJS.
